src/libraries/System.Net.Security/src/System/Net/Security/Pal.Android/SafeDeleteSslContext.cs

```diff
@@ -75,61 +89,100 @@ protected override void Dispose(bool disposing)
         [UnmanagedCallersOnly]
         private static unsafe void WriteToConnection(IntPtr connection, byte* data, int dataLength)
         {
-            SafeDeleteSslContext? context = (SafeDeleteSslContext?)GCHandle.FromIntPtr(connection).Target;
-            Debug.Assert(context != null);
+            GCHandle h = GCHandle.FromIntPtr(connection);
+            SafeDeleteSslContext? context = h.IsAllocated ? h.Target as SafeDeleteSslContext : null;
+            if (context is null || context._disposed)
+            {
+                Debug.Write("WriteToConnection: context is null");
+                return;
+            }
 
-            var inputBuffer = new ReadOnlySpan<byte>(data, dataLength);
+            try
+            {
+                lock (context._lock)
+                {
+                    var inputBuffer = new ReadOnlySpan<byte>(data, dataLength);
 
-            context._outputBuffer.EnsureAvailableSpace(dataLength);
-            inputBuffer.CopyTo(context._outputBuffer.AvailableSpan);
-            context._outputBuffer.Commit(dataLength);
+                    context._outputBuffer.EnsureAvailableSpace(dataLength);
+                    inputBuffer.CopyTo(context._outputBuffer.AvailableSpan);
+                    context._outputBuffer.Commit(dataLength);
+                }
+            }
+            catch (Exception ex)
+            {
+                Debug.Write("Exception Caught. - " + ex);
+            }
         }
 
         [UnmanagedCallersOnly]
         private static unsafe PAL_SSLStreamStatus ReadFromConnection(IntPtr connection, byte* data, int* dataLength)
         {
-            SafeDeleteSslContext? context = (SafeDeleteSslContext?)GCHandle.FromIntPtr(connection).Target;
-            Debug.Assert(context != null);
-
-            int toRead = *dataLength;
-            if (toRead == 0)
-                return PAL_SSLStreamStatus.OK;
-
-            if (context._inputBuffer.ActiveLength == 0)
+            GCHandle h = GCHandle.FromIntPtr(connection);
+            SafeDeleteSslContext? context = h.IsAllocated ? h.Target as SafeDeleteSslContext : null;
+            if (context is null || context._disposed)
```

Comment:
Same as above... this check seems ineffective since the Dispose may happen in between the check and the lock/usage below.